spec/services/json_validator_spec.rb

```diff
@@ -0,0 +1,517 @@
+require "spec_helper"
+
+RSpec.describe JsonValidator do
+  include FormResponseHelper
+
+  let(:data) { valid_data }
+  let(:schema_data) { File.read(Rails.root.join("config/schemas/form_response.json")) }
+  let(:schema) { JSON.parse schema_data }
+  subject { described_class.new(schema, data) }
+
+  describe "#invalid?" do
+    it "returns false if data is valid" do
+      expect(subject.invalid?).to be(false)
+    end
+
+    context "with invalid data" do
+      let(:data) { valid_data.except(:live_in_england) }
+
+      it "returns true" do
+        expect(subject.invalid?).to be(true)
+      end
+    end
+  end
+
+  describe "#errors" do
+    let(:error) { subject.errors.first }
+
+    it "returns no error if the data is valid" do
+      expect(subject.errors).to be_empty
+    end
+
+    context "when live_in_england missing" do
+      let(:data) { valid_data.except(:live_in_england) }
+
+      it "returns a missing keys error" do
+        expect(error["missing_keys"]).to contain_exactly("live_in_england")
+        expect(error["data"]).to be_nil
```

Comment:
Yes - there are a number of tests that look for `error["data"]` content. I added this one specially because the data content returned by JSONSchemer's `validate` method in this condition contains all the data submitted - so it important that it doesn't get logged. I'll split that out into a separate spec to make that clearer.